sw/device/lib/dif/dif_keymgr.h

```diff
@@ -0,0 +1,648 @@
+   * This is the last operational state of key manager. At this state, key
+   * manager can generate a versioned owner key or an owner identity seed that
+   * can be used to generate an owner identity using an asymmetric KDF.
+   */
+  kDifKeymgrStateOwnerRootKey,
+  /**
+   * Disabled state.
+   *
+   * This is a terminal state where key manager is no longer operational. At
+   * this state, the secret value of key manager is a random value.
+   */
+  kDifKeymgrStateDisabled,
+} dif_keymgr_state_t;
+
+/**
+ * Creates a new handle for key manager.
+ *
+ * This function does not actuate the hardware and must be called to initialize
+ * the handle that must be passed to other functions in this library in each
+ * boot stage. A typical usage of this library during different secure boot
+ * stages is as follows:
+ *
+ * - In Mask ROM:
+ *   - Create a new handle: `dif_keymgr_init()`.
+ *   - Initialize state: `dif_keymgr_initialize_state()`,
+ *   `dif_keymr_get_status()`.
+ *   - Advance state: `dif_keymgr_advance_state()`, `dif_keymgr_get_status()`.
+ * - In subsequent boot stages, i.e. ROM_EXT, BL0, kernel:
+ *   - Create a new handle: `dif_keymgr_init()`.
+ *   - Generate keys and/or identity seeds:
+ *     `dif_keymgr_generate_versioned_key()`,
+ *     `dif_keymgr_generate_identity_seed()`, `dif_keymgr_get_status()`.
+ *   - Read output: `dif_keymgr_read_output()`.
+ *   - Advance state: `dif_keymgr_advance_state()`, `dif_keymgr_get_status()`.
+ *
+ * @param params Hardware instantiation parameters.
+ * @param[out] keymgr Out-param for the initialized handle.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_keymgr_result_t dif_keymgr_init(dif_keymgr_params_t params,
+                                    dif_keymgr_t *keymgr);
+
+/**
+ * Initializes key manager after power-on reset.
+ *
+ * After PoR, key manager is in Reset state with a non-deterministic secret
+ * value. This function causes key manager to initialize its secret value using
+ * the random values generated by the entropy source and transition to
+ * Initialized state. Note that Initialized state is not an operational
+ * state and key manager state must be advanced one more time by calling
+ * `dif_keymgr_advance_state()` before keys or identity seeds can be generated.
+ *
+ * This is an asynchronous function because key manager state transitions
+ * involve KMAC operations that can take some time to complete. Clients must
+ * check the status of key manager using `dif_keymgr_get_status()` before
+ * calling other functions in this library.
+ *
+ * The entropy source must be initialized before this function is called because
+ * key manager relies on the entropy source to initialize its secret value.
+ *
+ * This function must be called only once per power cycle.
+ *
+ * @param keymgr A key manager handle.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_keymgr_result_t dif_keymgr_initialize_state(const dif_keymgr_t *keymgr);
```

Comment:
Thoughts: Is "initialize" a good word here? It feels confusable with "init". I don't know what the best word is, but I think something like "prime" or "seed" might be more appropriate.

Reply:
I had the same concern but thought prime/seed might be slightly inaccurate because this value is not used for computing the secret value of the following states.

This function can also be dropped but I am inclined to keep it until the HW design is finalized. Leaving this open for others to share their thoughts.